articles/introduction-to-dynamic-programming/index.md

```diff
@@ -0,0 +1,147 @@
+
+**It is not fast:** Iteration(using loops) is faster than recursion because every time a function is called, there is an overhead of allocating space for the function and all its data in the function stack. This causes a slight delay in recursive functions when compared to iteration. [This](https://stackoverflow.com/questions/10057443/explain-the-concept-of-a-stack-frame-in-a-nutshell) is a good read on the function call stack and how it works.
+
+In recursion, the same function can be called multiple times with the same arguments. In other words, the same result is calculated multiple times instead of just once. In dynamic programming, a recursive function is optimized by storing the intermediate results in a data structure. We store these results so that they are only calculated once. In other words, any recursive function in which the same results are calculated multiple times can be converted into DP. Let us look at the Fibonacci number example(covered in the previous article) to get a better idea.
+
+### Fibonacci Numbers
+```py
+def fibonacci(n):
+	if(n == 1 or n == 2):
+		return 1
+	return fibonacci(n - 1) + fibonacci(n - 2)
+n = int(input('Enter the nth term: '))
+print(fibonacci(n))
+```
+
+The above code is the recursive implementation of Fibonacci numbers covered in the previous article. The `fibonacci()` function is called twice: once to calculate `(n - 1)` and once to calculate `(n - 2)`. Therefore, the time complexity of this function is 2 raised to the power 'n'. $O(2^n)$. [This](https://www.geeksforgeeks.org/time-complexity-recursive-fibonacci-program/) is a good article on the time complexity of the recursive Fibonacci function. For example:
+
+Consider the case when n = 5.
+
+1. We call fibonacci(5), which inturn calls fibonacci(4) and fibonacci(3)
+
+2. fibonacci(4) inturn calls fibonacci(3) and fibonacci(2)
+
+3. fibonacci(3) inturn calls fibonacci(2) and fibonacci(1)
+
+4. fibonacci(2) and fibonacci(1) return 1, and the program terminates.
+
+As you can observe, fibonacci(4), fibonacci(3), fibonacci(2) are called multiple times. Take a look at the image below to get a better understanding:
+
+![Fibonacci](/engineering-education/introduction-to-dynamic-programming/fibonacci.png)
+[Image source](https://www.slitherintopython.com/assets/img/fibtree.png)
+
+A better way to solve this would be to store all the intermediate results to calculate them once. There are two ways to do this, but in this article, we will explore one method called tabulation. To understand the other method, check out [This](https://www.educative.io/courses/grokking-dynamic-programming-patterns-for-coding-interviews/m2G1pAq0OO0) article. In the tabulation method, we store the results of each function call in a data structure. Whenever an already stored value is needed, we fetch the value from the data structure instead of computing it repeatedly. For the Fibonacci example, we can use an array to store all the intermediate numbers.
```

Comment:
What is the other method? Briefly describe the method. Also, break the para into 2 smaller paras.